### PR TITLE
allow to use pages greater than total_pages

### PR DESCRIPTION
**Pull request by AlexKovalevych**

### CODE REVIEW

**Comment by drewolson** [comment]
I really struggle to understand the use case for this. Apologies, but I don't think this is something I want to include.

**Comment by AlexKovalevych** [comment]
what about this issue? https://github.com/drewolson/scrivener_ecto/issues/68

**Comment by AlexKovalevych** [comment]
and this one https://github.com/drewolson/scrivener_ecto/issues/69 ?

**Comment by drewolson** [comment]
The desire in both of these issues is a concrete way to know that you're on the last page. I'd accept a PR adding `last?` to [Scrivener.Page](https://github.com/drewolson/scrivener/blob/a2751b00d4d86c9e394934e6a2013b82fe122efc/lib/scrivener/page.ex) but accomplishing this by allowing pagination past the end of the collection feels counter-intuitive to me.

**Comment by speeddragon** [comment]
> ... but accomplishing this by allowing pagination past the end of the collection feels counter-intuitive to me.

I understand that it might be counter-intuitive to expect an empty response from a page that doesn't exist, but that is a common behaviour in several places (for example, databases with offset past the limit of existent rows).

But sending the first page result on a page that doesn't exist is even worse, IMHO. It would be nice if at least it would throw an error.

**Comment by AlexKovalevych** [comment]
> The desire in both of these issues is a concrete way to know that you're on the last page.
 
Well, my point is not to break backward compatibility. Since you changed the way library works for such cases, we simply can't use it anymore. It's probably fine for most of projects, but as i can see not for all. So, again i don't change the default behavior, just allow to be backward compatible.

**Comment by juanperi** [comment]
I'm also facing this. In my end, expected behavior for an out of bounds page would be to return 404. With the current implementation after migrating to version 2, i cannot rely anymore in no results.
I will have to check that the requested page is higher than the returned page.
Even adding `Page.last?` will force me to do so in every usage, checking page number against the params (and defaulting to 1). And also I'll have to parse the page number from the params as well, even knowing you do the same again inside the library

in a comment on issue #69 you answered:
> I'm not interested in allowing pagination beyond the end of the collection. See #71 for a proposed solution.

I don't think we are talking about allowing pagination beyond the end here, but allowing users of the library to detect when somebody tries to do so.

**Comment by drewolson** [comment]
After re-thinking this, the fact that it is guarded behind an option keyword seems reasonable to me. I'll reopen and merge.

File: lib/scrivener/paginater/ecto/query.ex [hunk 1]
```diff
@@ -17,7 +17,10 @@ defimpl Scrivener.Paginater, for: Ecto.Query do
       Keyword.get_lazy(options, :total_entries, fn -> total_entries(query, repo, caller) end)
 
     total_pages = total_pages(total_entries, page_size)
-    page_number = min(total_pages, page_number)
+    allow_out_of_range_pages = Keyword.get(options, :allow_out_of_range_pages, false)
```
Comment (drewolson) [inline lib/scrivener/paginater/ecto/query.ex]: Can we rename this to `allow_overflow_page_number` please?

**Review by drewolson — COMMENTED** [review]
One small comment, otherwise looks good.

**Review by drewolson — COMMENTED** [review]
One small comment, otherwise looks good.

File: lib/scrivener/paginater/ecto/query.ex (hunk 1 above)
Comment (AlexKovalevych) [reply lib/scrivener/paginater/ecto/query.ex]: Done

File: lib/scrivener/paginater/ecto/query.ex [hunk 2]
```diff
@@ -17,12 +17,20 @@ defimpl Scrivener.Paginater, for: Ecto.Query do
...
-    page_number = min(total_pages, page_number)
+    allow_overflow_page_number = Keyword.get(options, :allow_overflow_page_number, false)
+
+    page_number =
+      if allow_overflow_page_number, do: page_number, else: min(total_pages, page_number)
+
+    items =
+      if page_number > total_pages,
+        do: [],
+        else: entries(query, repo, page_number, page_size, caller, options)
```
Comment (drewolson) [inline lib/scrivener/paginater/ecto/query.ex]: Is there a reason to add this conditional? Efficiency?

Comment (AlexKovalevych) [reply lib/scrivener/paginater/ecto/query.ex]: yup, no need to make an extra db query in this case

Comment (drewolson) [reply lib/scrivener/paginater/ecto/query.ex]: Got it. In that case I'd prefer to inline this into the `entries` function itself, perhaps via guard clauses. It'll mean adding another argument to the method.

**Review by drewolson — APPROVED** [review]

**Comment by drewolson** [comment]
Thanks. I'll cut a new release later today or tomorrow.

**Pull request merged**